### PR TITLE
Allow strict mode enforcement of violations

### PR DESCRIPTION
**Pull request by alexevanczuk**
## What are you trying to accomplish?
More info here: https://github.com/Shopify/packwerk/discussions/241
TLDR: This enables `strict` mode for checkers so that packages can disallow new violations, even if they're listed in `package_todo.yml` files.

## What approach did you choose and why?
I chose to delegate the question to the checker as it's closely coupled to *where* a violation is being stored. That is, if package A has a new dependency violation on package B, we need to check package A `enforce_dependencies` to determine if the violation when listed violates strict mode. If package A has a new *privacy* violation on package B, we need to check the `enforce_privacy` property for package B. Since this is determined by the checker, it needs to be a checker interface method.

## What should reviewers focus on?
Let me know if there is a better way to implement this, or a better UX (e.g. the output, the API to turn it on, etc.). Also let me know if I'm missing any edge cases!

## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] It is safe to rollback this change.

### CODE REVIEW

File: lib/packwerk/checker.rb [hunk 1]
```diff
@@ -22,11 +22,28 @@ def included(base)
...
+      def find(violation_type)
+        checker_by_violation_type(violation_type)
+      end
+
+      private
+
+      sig { params(name: String).returns(Checker) }
+      def checker_by_violation_type(name)
+        @checker_by_violation_type ||= T.let(@checker_by_violation_type, T.nilable(T::Hash[String, T.nilable(Checker)]))
+        @checker_by_violation_type ||= Checker.all.map { |c| [c.violation_type, c] }.to_h
```
[the remark this answers is not in the document]
Comment (alexevanczuk) [reply lib/packwerk/checker.rb]: Sounds good

File: lib/packwerk/offense_collection.rb [hunk 2]
```diff
@@ -46,9 +50,15 @@ def add_offense(offense)
         @errors << offense
         return
       end
-      package_todo = package_todo_for(offense.reference.source_package)
-      unless package_todo.add_entries(offense.reference, offense.violation_type)
+
+      # require 'pry'
+      # binding.pry if offense.reference.constant.name.include?('Squid')
```
[the remark this answers is not in the document]
Comment (alexevanczuk) [reply lib/packwerk/offense_collection.rb]: Whoops, yep!

File: lib/packwerk/checker.rb [hunk 3]
```diff
@@ -22,11 +22,29 @@ def included(base)
...
+      sig { params(violation_type: String).returns(Checker) }
+      def find(violation_type)
+        checker_by_violation_type(violation_type)
+      end
+
+      private
+
+      sig { params(name: String).returns(Checker) }
+      def checker_by_violation_type(name)
+        @checker_by_violation_type ||= T.let(Checker.all.map do |c|
```
Comment (gmcgibbon) [inline lib/packwerk/checker.rb:35]:
```suggestion
        @checker_by_violation_type ||= T.let(Checker.all.to_h do |checker|
```
You can use Array#to_h to map and cast in one call.

File: test/unit/cli_test.rb [hunk 4]
```diff
@@ -143,6 +143,10 @@ def show_stale_violations(_offense_collection, _fileset)
         def identifier
           "custom"
         end
+
+        def show_strict_mode_violations(offenses)
```
Comment (gmcgibbon) [inline test/unit/cli_test.rb:147]:
```suggestion
        def show_strict_mode_violations(_offenses)
```

File: lib/packwerk/package.rb [hunk 5]
```diff
@@ -30,7 +30,12 @@ def initialize(name:, config:)
 
     sig { returns(T::Boolean) }
     def enforce_dependencies?
-      @config["enforce_dependencies"] == true
+      [true, "strict"].include?(@config["enforce_dependencies"])
+    end
+
+    sig { returns(T::Boolean) }
+    def strictly_enforce_dependencies?
```
Comment (gmcgibbon) [inline lib/packwerk/package.rb:37]: Is this used?

File: lib/packwerk/offense_collection.rb [hunk 6]
```diff
@@ -46,9 +50,13 @@ def add_offense(offense)
         @errors << offense
         return
       end
-      package_todo = package_todo_for(offense.reference.package)
-      unless package_todo.add_entries(offense.reference, offense.violation_type)
+
+      already_listed = package_todo_for(offense.reference.package).add_entries(offense.reference,
+        offense.violation_type)
+      if !already_listed
```
Comment (gmcgibbon) [inline lib/packwerk/offense_collection.rb:56]: Let's make this a private method too for readability.

File: test/unit/checker_test.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# typed: true
+# frozen_string_literal: true
+
+require "test_helper"
+
+module Packwerk
+  class CheckerTest < Minitest::Test
+    include FactoryHelper
```
Comment (gmcgibbon) [inline test/unit/checker_test.rb:8]:
```suggestion
```

File: test/unit/checker_test.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+    test "#find is correctly able to find the right checker" do
+      # Load in the checker so its registered
+      require "packwerk/reference_checking/checkers/dependency_checker"
```
Comment (gmcgibbon) [inline test/unit/checker_test.rb:12]: I assume https://github.com/Shopify/packwerk/blob/72aa3cf9dd1c4035356e5474e115583bada071da/lib/packwerk/run_context.rb#L12 is implicitly loading this in non-test runs. Should we just add a require to the main entrypoint like we did with the offense formatter?

Comment (alexevanczuk) [reply test/unit/checker_test.rb:11]: Yep, makes a lot of sense. This was also able to remove the unused `DEFAULT_CHECKERS` param which was just implicitly used to register the constant.

File: lib/packwerk/package.rb (hunk 5 above)
Comment (alexevanczuk) [reply lib/packwerk/package.rb:37]: Nope! An artifact of a spike that I didn't end up using, removing.

File: lib/packwerk/offense_collection.rb (hunk 6 above)
Comment (alexevanczuk) [reply lib/packwerk/offense_collection.rb:56]: sounds good

**Comment by alexevanczuk** [comment]
Incorporated the newest round of feedback @gmcgibbon!

**Review by gmcgibbon — APPROVED** [review]
